completion_aggregator/transformers.py

```diff
@@ -79,3 +84,119 @@ def transform(self, usage_info, block_structure):  # pylint: disable=unused-argu
             if completion_mode != XBlockCompletionMode.EXCLUDED:
                 aggregators = self.calculate_aggregators(block_structure, block_key)
                 block_structure.set_transformer_block_field(block_key, self, self.AGGREGATORS, aggregators)
+
+
+class BaseAggregatorXApiTransformer(XApiTransformer):
+    """
+    Base transformer for all completion aggregator events.
+    """
+
+    object_type = None
+
+    def get_object(self) -> Activity:
+        """
+        Get object for xAPI transformed event.
+        """
+        if not self.object_type:
+            raise NotImplementedError()  # pragma: no cover
+
+        return Activity(
+            id=self.get_object_iri("xblock", self.get_data("data.block_id")),
+            definition=ActivityDefinition(
+                type=self.object_type,
+            ),
+        )
+
+
+class BaseProgressTransformer(BaseAggregatorXApiTransformer):
+    """
+    Base transformer for completion aggregator progress events.
+    """
+
+    _verb = Verb(
+        id=constants.XAPI_VERB_PROGRESSED,
+        display=LanguageMap({constants.EN: constants.PROGRESSED}),
+    )
+    object_type = None
+    additional_fields = ('result', )
+
+    @openedx_filter(
+        filter_type="completion_aggregator.xapi.base_progress.get_object",
+    )
+    def get_object(self) -> Activity:
+        """
+        Get object for xAPI transformed event.
+        """
+        return super().get_object()
+
+    def get_result(self) -> Result:
+        """
+        Get result for xAPI transformed event.
+        """
+        return Result(
+            completion=self.get_data("data.percent") == 1.0,
+            score={
+                "scaled": self.get_data("data.percent") or 0
```

Comment:
Why do we call it a "scaled" score?

Reply:
Huh.. I started to answer you by pointing to the [emitted grading events](https://github.com/openedx/event-routing-backends/blob/980b2aa264261430c5b8dfaeea70c02d44e5e08a/event_routing_backends/processors/xapi/event_transformers/grading_events.py#L63), but since this actually a progress % not a score, this line is actually a misuse of the [`tincan.Result.score` attribute](http://rusticisoftware.github.io/TinCanJS/doc/api/latest/classes/TinCan.Score.html).

So I've updated it to use an Extension instead.